pkg/pool-utils/contracts/RebalancedBasePool.sol

```diff
@@ -0,0 +1,128 @@
+            userData
+        );
+        _ensureAssetManagerBalanced(poolId, amountsIn, dueProtocolFeeAmounts, true);
+        return (amountsIn, dueProtocolFeeAmounts);
+    }
+
+    function onExitPool(
+        bytes32 poolId,
+        address sender,
+        address recipient,
+        uint256[] memory balances,
+        uint256 lastChangeBlock,
+        uint256 protocolSwapFeePercentage,
+        bytes memory userData
+    ) public virtual override returns (uint256[] memory amountsOut, uint256[] memory dueProtocolFeeAmounts) {
+        (amountsOut, dueProtocolFeeAmounts) = super.onExitPool(
+            poolId,
+            sender,
+            recipient,
+            balances,
+            lastChangeBlock,
+            protocolSwapFeePercentage,
+            userData
+        );
+        _ensureAssetManagerBalanced(poolId, amountsOut, dueProtocolFeeAmounts, false);
+        return (amountsOut, dueProtocolFeeAmounts);
+    }
+
+    function _ensureAssetManagerBalanced(
+        bytes32 poolId,
+        uint256[] memory amounts,
+        uint256[] memory protocolFeeAmounts,
+        bool positive
+    ) internal view {
+        uint256 totalTokens = _getTotalTokens();
+        if (totalTokens == 0) return;
+        _ensureAssetManagerBalanced(poolId, _token0, amounts[0], protocolFeeAmounts[0], positive);
+        if (totalTokens == 1) return;
+        _ensureAssetManagerBalanced(poolId, _token1, amounts[1], protocolFeeAmounts[1], positive);
+        if (totalTokens == 2) return;
+        _ensureAssetManagerBalanced(poolId, _token2, amounts[2], protocolFeeAmounts[2], positive);
+        if (totalTokens == 3) return;
+        _ensureAssetManagerBalanced(poolId, _token3, amounts[3], protocolFeeAmounts[3], positive);
+        if (totalTokens == 4) return;
+        _ensureAssetManagerBalanced(poolId, _token4, amounts[4], protocolFeeAmounts[4], positive);
+        if (totalTokens == 5) return;
+        _ensureAssetManagerBalanced(poolId, _token5, amounts[5], protocolFeeAmounts[5], positive);
+        if (totalTokens == 6) return;
+        _ensureAssetManagerBalanced(poolId, _token6, amounts[6], protocolFeeAmounts[6], positive);
+        if (totalTokens == 7) return;
+        _ensureAssetManagerBalanced(poolId, _token7, amounts[7], protocolFeeAmounts[7], positive);
+    }
+
+    function _ensureAssetManagerBalanced(
+        bytes32 poolId,
+        IERC20 token,
+        uint256 amount,
+        uint256 protocolFeeAmount,
+        bool positive
+    ) internal view {
+        (uint256 cash, uint256 managed, , address assetManager) = getVault().getPoolTokenInfo(poolId, token);
+        if (assetManager != address(0)) {
+            uint256 finalCash = (positive ? cash.add(amount) : cash.sub(amount)).sub(protocolFeeAmount);
+            bool shouldRebalance = IAssetManager(assetManager).shouldRebalance(finalCash, managed);
+            _require(!shouldRebalance || _relayer.hasCalledPool(poolId), Errors.BASE_POOL_RELAYER_NOT_CALLED);
```

Comment:
While this is correct, it is currently lacking short-circuiting behavior: even after we find an asset manager requires rebalance and we check that the relayer has been called, we continue looking at all others. 

Perhaps this function could return `true` if the AM needs rebalancing and the relayer has been called? And then the version that calls this would immediately exit and skip all other AMs.